### ISSUE TITLE
Question Regarding View Selection in the Test Dataloader

### ISSUE THREAD

**Issue by XinhanWang** [yangaonidaye123/NRCL#2]
Hi there!

I noticed that in the following repositories:

[Cross-Modal-Center-Loss](https://github.com/LongLong-Jing/Cross-Modal-Center-Loss/blob/5025837561a5e059940462edde99aafe80680871/tools/test_dataloader.py#L83-L93)
[RONO](https://github.com/penghu-cs/RONO/blob/978d92887ebb8bbb36a240a6e019d4f4902869d4/tools/test_two_feature_one_raw_dataloader.py#L88-L100)
[DAC](https://github.com/ganchaofan0000/DAC/blob/370506570e285833149448d910e3801db185bf82/tools/dataset.py#L204-L219)
and your NRCL
the test dataloader randomly selects a view when multiple image views are available. While this randomness might help improve robustness during training, it introduces unnecessary variance during evaluation, leading to unstable and less reproducible test results.
https://github.com/yangaonidaye123/NRCL/blob/bf4b4f339472d3ca200004ca9e02882b6271319b/tools/dataset.py#L202-L218
To ensure consistent and reliable evaluation, would it be better to fix the view selection at test time (e.g., fix the random seed)?

I’d appreciate your thoughts on this!

**Comment by yangaonidaye123** [comment]
Thanks very much for your attention!

Although we have already set random.seed(seed) at the program entry to keep the random number generation consistent, your suggestion is indeed valuable. The current strategy randomly selects views when multiple image views are available, which may still introduce unnecessary evaluation variance even with a fixed global seed.

We appreciate this insight. We will follow your suggestion and run additional experiments by fixing the view selection during evaluation to verify its impact on stability and reproducibility. You are also welcome to test both settings to compare the consistency of the results.

**Comment by XinhanWang** [comment]
Hi author,
Thank you for this great work.
I would like to suggest a modification to the data loading process to ensure consistency in view selection during the evaluation/testing phase. Currently, if the test views are selected randomly, the evaluation results might fluctuate simply due to different view sampling, making it harder to reproduce specific results or compare epochs fairly.
I modified the code to use a torch.Generator with a fixed seed based on the item index during the non-training phase. This ensures that every time we valid/test, the same object always uses the same set of views.
Here is the implementation:
```python
img_list = []
img_index_list = []

# Logic for view selection
if self.partition == "train":
    num_view = 2
else:
    num_view = self.test_view_num
    # Ensure deterministic view selection for validation/test
    if not self.random_test_view:
        g = torch.Generator()
        g.manual_seed(item + self.random_seed)

for view in range(num_view):
    if self.partition == "train" or self.random_test_view:
        # Keep random strategy for training or if explicitly requested
        img_idx = torch.randint(0, 180, (1,)).item()
        while img_idx in img_index_list:
            img_idx = torch.randint(0, 180, (1,)).item()
    else:
        # Use fixed generator for reproducible testing
        img_idx = torch.randint(0, 180, (1,), generator=g).item()
        while img_idx in img_index_list:
            img_idx = torch.randint(0, 180, (1,), generator=g).item()
            
    img_index_list.append(img_idx)
    
    # ... (rest of the path loading logic) ...
    if self.dataset == 'ModelNet40':
        img_names = self.dataset_dir+'ModelNet40-Images-180/%s/%s/%s.%d.png' % (names[0], names[1][:-4], names[1][:-4], img_idx)
    else:
        img_names = self.dataset_dir+'ModelNet10-Images-180/%s/%s/%s.%d.png' % (names[0], names[1][:-4], names[1][:-4], img_idx)
    
    img = Image.open(img_names).convert('RGB')
    img_list.append(img)
```
Observation:
I noticed that when using this fixed view strategy, the mAP dropped slightly compared to the fully random selection. This might be because random sampling occasionally picks "easier" or more discriminative views by chance, whereas a fixed seed might lock in some "harder" views. However, I believe consistency is more important for fair comparison.
Hope this helps!
